core/headerchain.go

```diff
@@ -138,25 +138,27 @@ func (hc *HeaderChain) getFinalizedNumber(header *types.Header) uint64 {
 	return 0
 }
 
-// isFinalizedBlockHigher returns true when the new block's finalized block is higher than current block.
-func (hc *HeaderChain) isFinalizedBlockHigher(header *types.Header, curHeader *types.Header) bool {
+// reorgNeededWithFastFinality returns true when the finalized block is higher, otherwise backoff to compare the td.
+func (hc *HeaderChain) reorgNeededWithFastFinality(forker *ForkChoice, current *types.Header, header *types.Header) (bool, error) {
 	p, ok := hc.engine.(consensus.PoSA)
 	if !ok {
-		return false
+		return forker.ReorgNeeded(current, header)
 	}
 
-	ancestor := rawdb.FindCommonAncestor(hc.chainDb, header, curHeader)
+	ancestor := rawdb.FindCommonAncestor(hc.chainDb, current, header)
 	if ancestor == nil {
-		return false
+		return forker.ReorgNeeded(current, header)
 	}
 
 	finalized := p.GetFinalizedHeader(hc, header, header.Number.Uint64()-ancestor.Number.Uint64())
-	curFinalized := p.GetFinalizedHeader(hc, curHeader, curHeader.Number.Uint64()-ancestor.Number.Uint64())
-	if finalized == nil || curFinalized == nil {
-		return false
+	curFinalized := p.GetFinalizedHeader(hc, current, current.Number.Uint64()-ancestor.Number.Uint64())
+	if finalized == nil || curFinalized == nil || finalized.Number.Uint64() == curFinalized.Number.Uint64() {
+		return forker.ReorgNeeded(current, header)
+	} else if finalized.Number.Uint64() > curFinalized.Number.Uint64() {
+		return true, nil
```

Comment:
if(curFinalized.Number.Uint64()> ancestor.Number.Uint64()){
    we can log that conflicting blocks have been finalized here.
}